v2/tests/git_initializer_test.go

```diff
@@ -179,46 +231,150 @@ func TestMain(t *testing.T) {
+		require.NoError(t, err, "error getting event")
+
+		phase := event.Worker.Status.Phase
+		if phase.IsTerminal() {
+			require.Equal(t, tc.expectedWorkerPhase, phase, "worker's terminal phase does not match expected")
+			break
+		}
+	}
+}
+
+func assertWorkerLogs(
+	t *testing.T,
+	ctx context.Context,
+	client sdk.APIClient,
+	e core.Event,
+	tc testcase) {
+
+	// Check vcs container logs
+	selector := &core.LogsSelector{Container: "vcs"}
+	logs := captureLogs(t, ctx, client, e, selector)
+
+	if tc.expectedVCSLogsContains != "" {
+		require.Contains(t, logs, tc.expectedVCSLogsContains, "vcs init container logs do not match expected")
+	}
+
+	// We expect the vcs init container to fail for this test,
+	// therefore, assertions on worker logs are not applicable
+	if tc.name == "GitHub - vcs failure" {
+		return
+	}
+
+	// Check worker container logs
+	selector = &core.LogsSelector{}
+	logs = captureLogs(t, ctx, client, e, selector)
+
+	if tc.expectedWorkerLogsContains != "" {
+		require.Contains(t, logs, tc.expectedWorkerLogsContains, "worker's logs do not match expected")
+	} else {
+		// Look for default logs that we expect must exist
+		require.Contains(t, logs, "brigade-worker version", "worker's logs do not match expected")
+	}
+}
+
+func assertJobTerminalPhase(
+	t *testing.T,
+	ctx context.Context,
+	client sdk.APIClient,
+	e core.Event,
+	tc testcase) {
+
+	for {
+		event, err := client.Core().Events().Get(ctx, e.ID)
+		require.NoError(t, err, "error getting event")
+
+		jobs := event.Worker.Jobs
+		require.Equal(t, 1, len(jobs), "expected job count is not 1")
+
+		job := jobs["ls"]
+		require.NotNil(t, job, "expected job does not exist")
+
+		phase := job.Status.Phase
+		if phase.IsTerminal() {
+			require.Equal(t, tc.expectedJobPhase, phase, "job's terminal phase does not match expected")
+			break
+		}
+	}
+}
+
+func assertJobLogs(
+	t *testing.T,
+	ctx context.Context,
+	client sdk.APIClient,
+	e core.Event,
+	tc testcase) {
+
+	selector := &core.LogsSelector{Job: "ls"}
+	logs := captureLogs(t, ctx, client, e, selector)
+
+	if tc.expectedJobLogsContains != "" {
+		require.Contains(t, logs, tc.expectedJobLogsContains, "job's logs do not match expected")
+	}
+}
+
+func captureLogs(
+	t *testing.T,
+	ctx context.Context,
+	client sdk.APIClient,
+	e core.Event,
+	selector *core.LogsSelector) string {
+
+	logEntryCh, errCh, err :=
+		client.Core().Events().Logs().Stream(
+			ctx,
+			e.ID,
+			selector,
+			&core.LogStreamOptions{},
+		)
+	require.NoError(t, err, "error acquiring log stream")
+
+	var b bytes.Buffer
+	for {
+		select {
+		case logEntry, ok := <-logEntryCh:
+			if ok {
+				b.Write([]byte(logEntry.Message))
+			} else {
+				logEntryCh = nil
+			}
+		case err, ok := <-errCh:
+			if ok {
+				t.Fatalf("error from log stream: %s", err)
+			}
+			errCh = nil
+		case <-ctx.Done():
+			break
+		}
+
+		// log and err channels empty, let's return logs
+		if logEntryCh == nil && errCh == nil {
+			return b.String()
+		}
+	}
+}
```

Comment:
If the container you're capturing logs from produces enough lines of output, it's possible to blow up the heap here.

fwiw, this is why the API always streams logs to begin with and doesn't expose any option to retrieve them as a single chunk-- risk of blowing up the heap on either the client side _or_ server side.

Instead of capturing a potentially large chunk of logs, which I believe you are then asserting contain some text, what if you pass in the text you're expecting to find, ditch the buffer, and scan the log stream for the specified text until you either find it or reach the end of the stream?

Reply:
> If the container you're capturing logs from produces enough lines of output, it's possible to blow up the heap here.

And I acknowledge that as written, these tests probably won't do that.

I'm more concerned about:

1. What happens when something goes wrong. We don't want an OOM obscuring the actual issue.

1. Keeping this flexible enough to accommodate hypothetical test cases that do produce larger amounts of logs.

Reply:
Good call; updated!